charmcraft/commands/store/__init__.py

```diff
@@ -722,6 +737,222 @@ def run(self, parsed_args):
+        target_channel = ChannelData.from_str(parsed_args.target_channel)
+
+        if target_channel == source_channel:
+            raise CraftError("Cannot promote from a channel to the same channel.")
+        if target_channel.risk > source_channel.risk:
+            command_parts = [
+                "charmcraft",
+                "promote-bundle",
+                target_channel.name,
+                source_channel.name,
+            ]
+            if parsed_args.output_bundle:
+                command_parts.extend(["--output-bundle", parsed_args.output_bundle])
+            for exclusion in parsed_args.exclude:
+                command_parts.extend(["--exclude"], exclusion)
+            command = " ".join(command_parts)
+            raise CraftError(
+                f"Target channel ({target_channel.name}) must be lower risk "
+                f"than the source channel ({source_channel.name}).\n"
+                f"Did you mean: {command}"
+            )
+        if target_channel.track != source_channel.track:
+            emit.message(
+                "Promoting to a different track (from "
+                f"{source_channel.track} to {target_channel.track})"
+            )
+
+        output_bundle: typing.Optional[pathlib.Path] = parsed_args.output_bundle
+        if output_bundle is not None and output_bundle.exists():
+            if output_bundle.is_file() or output_bundle.is_symlink():
+                emit.verbose(f"Overwriting existing bundle file: {str(output_bundle)}")
+            elif output_bundle.is_dir():
+                emit.debug(f"Creating bundle file in {str(output_bundle)}")
+                output_bundle /= "bundle.yaml"
+            else:
+                raise CraftError(f"Not a valid bundle output path: {str(output_bundle)}")
+        elif output_bundle is not None:
+            if not output_bundle.suffix:
+                output_bundle /= "bundle.yaml"
+            for parent in output_bundle.parents:
+                if parent.exists():
+                    if os.access(parent, os.W_OK):
+                        break
+                    raise CraftError(f"Bundle output directory not writable: {str(parent)}")
+
+        # Load bundle
+        bundle_path = self.config.project.dirpath / "bundle.yaml"
+        bundle_config = load_yaml(bundle_path)
+        if bundle_config is None:
+            raise CraftError(f"Missing or invalid main bundle file: {(str(bundle_path))}")
+        bundle_name = bundle_config.get("name")
+        if not bundle_name:
+            raise CraftError(
+                "Invalid bundle config; missing a 'name' field indicating the bundle's name in "
+                f"file {str(bundle_path)!r}."
+            )
+        emit.progress("Determining charms to promote")
+        charms = [c["charm"] for c in bundle_config["applications"].values()]
+        errant_excludes = []
+        for excluded in parsed_args.exclude:
+            try:
+                charms.remove(excluded)
```

Comment:
Instead of creating a full list of charms and then removing the excluded ones, wouldn't it be better to create the initial list without the excluded items?

Reply:
The best I can come up with without sacrificing the error checking is:

```python
        charms = []
        excluded_charms = set()
        for application in bundle_config["applications"].values():
            if application["charm"] in parsed_args.exclude:
                excluded_charms.add(application["charm"])
            else:
                charms.append(application["charm"])
        errant_excludes = set(parsed_args.exclude) - excluded_charms
```

I'm not sure it gains us anything, but I'm happy to put it (or a modification) in instead since I don't see it losing us anything either.

Reply:
[resolved after direct conversation]